### PR TITLE
Avoid metadata import/export in delta backups. (fix vatesfr/xo-web#651)

### PR DESCRIPTION
**Pull request by Wescoeur**

### CODE REVIEW

**Comment by olivierlambert** [comment]
@julien-f tested and it works. If it's OK for you, you can merge!

File: src/xo-mixins/backups.js [hunk 1]
```diff
@@ -63,21 +63,26 @@ export default class {
...
+
     for (const deltaDir of deltaDirs) {
       const files = await readdir(`${path}/${deltaDir}`)
-      const deltaBackups = filter(files, xvaFilter)
+      const deltaBackups = filter(files, deltaBackupFilter)
 
       backups.push(...mapToArray(
         deltaBackups,
-        deltaBackup => `${deltaDir}/${deltaBackup}`
+        deltaBackup => {
```
Comment (julien-f) [inline src/xo-mixins/backups.js]: Use slice here to remove the suffix.
```js
const SUFFIX = '.json'
const SUFFIX_LENGTH = SUFFIX.length
```

File: src/xo-mixins/backups.js [hunk 2]
```diff
@@ -377,6 +455,7 @@ export default class {
     await ensureDir(path)
 
     const info = {
+      version: '1.0.0',
       vbds: [],
```
Comment (julien-f) [inline src/xo-mixins/backups.js]: Index VBDs by their refs like VDIs.

File: src/xo-mixins/backups.js [hunk 3]
```diff
@@ -384,40 +463,35 @@ export default class {
...
-      if (!vbd.VDI) {
+    for (const vbd of vm.$VBDs) {
+      const vdiId = vbd.VDI
+
+      if (!vdiId || vbd.type !== 'Disk') {
         continue
       }
 
-      if (vbd.is_cd_drive) {
+      info.vbds.push({ ...vbd })
```
Comment (julien-f) [inline src/xo-mixins/backups.js]: No need to clone the object.

File: src/xo-mixins/backups.js [hunk 4]
```diff
@@ -384,40 +463,35 @@ export default class {
...
-      const vdiXo = this._xo.getObject(vbd.VDI)
-      const vdi = xapi.getObject(vdiXo._xapiId)
-      const vdiUUID = vdi.uuid
+      const vdi = vbd.$VDI
 
-      info.vbds.push({
-        ...xapi.getObject(vbd._xapiId),
-        xoVdi: vdiUUID
-      })
+      info.vdis[vdiId] = { ...vdi }
```
Comment (julien-f) [inline src/xo-mixins/backups.js]: Idem.

File: src/xo-mixins/backups.js [hunk 5]
```diff
@@ -441,21 +515,24 @@ export default class {
...
-        writeFile(infoPath, JSON.stringify(info), {flag: 'wx'})
-      ])
+      const vifs = info.vifs = {}
+
+      for (const vif of vm.$VIFs) {
+        vifs[vif.$ref] = vif
+      }
+
+      info.vm = {
+        ...vm
```
Comment (julien-f) [inline src/xo-mixins/backups.js]: No need to clone.

File: src/xo-mixins/backups.js [hunk 6]
```diff
@@ -472,7 +549,25 @@ export default class {
...
+      await Promise.all(
+        mapToArray(backups.slice(0, nOldBackups), async backup => {
+          // Remove json file.
+          await unlink(`${path}/${backup}`)
+
+          const xva = /^(.+)\.json$/.exec(backup)[1]
+
+          // Remove xva file.
+          // Version 0.0.0 (Legacy) Delta Backup.
+          if (xva) {
```
Comment (julien-f) [inline src/xo-mixins/backups.js]: This condition does not mean anything.

File: src/xo-mixins/backups.js [hunk 7]
```diff
@@ -472,7 +549,25 @@ export default class {
...
+    const backups = await this._listDeltaVmBackups(path)
+    const nOldBackups = backups.length - depth
+
+    if (nOldBackups > 0) {
+      await Promise.all(
+        mapToArray(backups.slice(0, nOldBackups), async backup => {
+          // Remove json file.
+          await unlink(`${path}/${backup}`)
+
+          const xva = /^(.+)\.json$/.exec(backup)[1]
```
Comment (julien-f) [inline src/xo-mixins/backups.js]: Use this regexp.

File: src/xo-mixins/backups.js [hunk 8]
```diff
@@ -489,76 +584,42 @@ export default class {
...
+    let vm
 
-    // Import VDIs.
-    const vdiIds = {}
-    await Promise.all(
-      mapToArray(
-        info.vdis,
-        async vdiInfo => {
-          vdiInfo.sr = sr._xapiId
+    if (delta.version === '1.0.0') {
```
Comment (julien-f) [inline src/xo-mixins/backups.js]: Use [`semver.satisfies(delta.version, '^1')`](https://www.npmjs.com/package/semver).

File: src/remote-handlers/local.js [hunk 9]
```diff
@@ -63,7 +64,19 @@ export default class LocalHandler extends RemoteHandlerAbstract {
...
+      stream = await fs.createReadStream(this._getFilePath(file), options)
+      await eventToPromise(stream, 'readable')
+    } catch (error) {
+      if (error.code === 'ENOENT') {
+        throw new Error(`${file} not found in this remote.`)
+      }
+      throw error
+    }
+
+    return stream
```
Comment (julien-f) [inline src/remote-handlers/local.js]: Try to keep it as simple as possible, do not use `try {} catch (error) {}`.

Comment (julien-f) [inline src/remote-handlers/local.js]: And move in `RemoteHandlerAbstract`.

File: src/remote-handlers/abstract.js [hunk 10]
```diff
@@ -76,11 +76,14 @@ export default class RemoteHandlerAbstract {
 
   async createReadStream (file, options) {
     const stream = await this._createReadStream(file)
+    await eventToPromise(stream, 'readable')
```
Comment (julien-f) [inline src/remote-handlers/abstract.js]: Add a `FIXME: the readable event may have already been emitted` .

File: src/remote-handlers/abstract.js [hunk 11]
```diff
@@ -76,11 +76,14 @@ export default class RemoteHandlerAbstract {
...
+
     if (stream.length === undefined) {
       try {
         stream.length = await this.getSize(file)
       } catch (_) {}
     }
```
Comment (julien-f) [inline src/remote-handlers/abstract.js]: It can be done in parallel of waiting for `readable`.

File: src/remote-handlers/local.js [hunk 12]
```diff
@@ -80,5 +80,4 @@ export default class LocalHandler extends RemoteHandlerAbstract {
     const stats = await fs.stat(this._getFilePath(file))
     return stats.size
   }
-
```
Comment (julien-f) [inline src/remote-handlers/local.js]: Change unrelated.

File: src/xo-mixins/backups.js [hunk 13]
```diff
@@ -40,6 +43,9 @@ const getVdiTimestamp = name => {
   return arr[1] || undefined
 }
 
+const getDeltaBackupNameWithoutExt = name =>
+  name.slice(0, name.length - DELTA_BACKUP_EXT_LENGTH)
```
Comment (julien-f) [inline src/xo-mixins/backups.js]: No need to explicit `name.length`.

File: src/xo-mixins/backups.js [hunk 14]
```diff
@@ -51,51 +57,34 @@ export default class {
     const handler = await this._xo.getRemoteHandler(remoteId)
 
     // List backups. (Except delta backups)
-    const xvaFilter = file => endsWith(file, '.xva')
+    const backupFilter = file => endsWith(file, '.xva')
```
Comment (julien-f) [inline src/xo-mixins/backups.js]: Mutualize and rename `isDeltaBackup(file)`.

File: src/xo-mixins/backups.js [hunk 15]
```diff
@@ -344,19 +398,14 @@ export default class {
...
-    }
+    return backups.slice(j, i + 1)
   }
 
   // -----------------------------------------------------------------
 
   async _listDeltaVmBackups (handler, dir) {
     const files = await handler.list(dir)
-    return await sortBy(filter(files, (fileName) => /^\d+T\d+Z_.*\.(?:xva|json)$/.test(fileName)))
+    return await sortBy(filter(files, (fileName) => /^\d+T\d+Z_.*\.json$/.test(fileName)))
```
Comment (julien-f) [inline src/xo-mixins/backups.js]: Use `isDeltaBackup(file)`.

File: src/xo-mixins/backups.js [hunk 16]
```diff
@@ -452,21 +497,22 @@ export default class {
...
-    const xvaPath = `${dir}/${backupFormat}.xva`
-    const infoPath = `${dir}/${backupFormat}.json`
+    const infoPath = `${dir}/${backupFormat}${DELTA_BACKUP_EXT}`
 
     try {
-      await Promise.all([
-        this._backupVm(vm, handler, xvaPath, {onlyMetadata: true}),
-        handler.outputFile(infoPath, JSON.stringify(info), {flag: 'wx'})
-      ])
+      const vifs = info.vifs = {}
```
Comment (julien-f) [inline src/xo-mixins/backups.js]: Uniformize with VBDs and VDIs.

File: src/xo-mixins/backups.js [hunk 17]
```diff
@@ -452,21 +497,22 @@ export default class {
...
+
+      for (const vif of vm.$VIFs) {
+        vifs[vif.$ref] = vif
+      }
+
+      info.vm = vm
+
+      await handler.outputFile(infoPath, JSON.stringify(info), {flag: 'wx'})
```
Comment (julien-f) [inline src/xo-mixins/backups.js]: You could format the JSON file nicely (`JSON.stringify(data, null, 2)`).

File: src/xo-mixins/backups.js [hunk 18]
```diff
@@ -500,76 +560,43 @@ export default class {
...
-    const vmName = vm.name_label
+    const delta = JSON.parse(await handler.readFile(`${filePath}${DELTA_BACKUP_EXT}`))
+    let vm
 
-    // Disable start and change the VM name label during import.
-    await Promise.all([
-      xapi.addForbiddenOperationToVm(vm.$id, 'start', 'Delta backup import...'),
-      xapi._setObjectProperties(vm, { name_label: `[Importing...] ${vmName}` })
-    ])
+    if (!delta.version) {
```
Comment (julien-f) [inline src/xo-mixins/backups.js]: Test that version satisfies `^1` in `Xapi#importDeltaVm()` and update `Xapi#exportDeltaVm()` consequently.

File: src/xo-mixins/backups.js [hunk 19]
```diff
@@ -344,19 +398,14 @@ export default class {
...
-    }
+    return backups.slice(j, i + 1)
   }
 
   // -----------------------------------------------------------------
 
   async _listDeltaVmBackups (handler, dir) {
     const files = await handler.list(dir)
-    return await sortBy(filter(files, (fileName) => /^\d+T\d+Z_.*\.(?:xva|json)$/.test(fileName)))
+    return await sortBy(filter(files, (fileName) => isDeltaBackup(fileName)))
```
Comment (julien-f) [inline src/xo-mixins/backups.js]: Do not use an intermediary function.

**Pull request merged**